client/lib/protect-form/index.tsx

```diff
@@ -16,10 +12,10 @@ import { Subtract } from 'utility-types';
  * Module variables
  */
 const debug = debugModule( 'calypso:protect-form' );
-let formsChanged = [];
+let formsChanged: any[] = [];
```

Comment:
Appropriate use of explicit any?

Reply:
`formsChanged` is an array of React component instances. That's probably an instance of the `React.Component` class?

There are two component classes whose instances can end up in `formsChanged`:
- component wrapped with the `protectForm` HOC
- instance of the `ProtectFormGuard` component

Both variants do the same job, but the former is imperative (the wrapped comonent must call `markSaved` and `markChanged` explicitly) and the latter is declarative (just mounting `<ProtectFormGuard>` marks the form as changed, unmounting marks as saved).

Reply:
I understand what the expected contents of the array are, but I'm having a lot of difficulty expressing their types 😞

Reply:
I pushed a commit where I'm getting stuck trying to type this:

The `ProtectFormGuard` part of this union seems to work well

https://github.com/Automattic/wp-calypso/blob/74d9b8d4eca31c5a78b9246856e30ad2a9a0de8c/client/lib/protect-form/index.tsx#L16

but the `ReturnType< typeof protectForm >` isn't working, these `this` don't match the type as I'd hoped:

https://github.com/Automattic/wp-calypso/blob/74d9b8d4eca31c5a78b9246856e30ad2a9a0de8c/client/lib/protect-form/index.tsx#L67-L68

Reply:
```
client/lib/protect-form/index.tsx:67:36 - error TS2345: Argument of type 'this' is not assignable to parameter of type 'ProtectableForm'.
  Type 'this' is not assignable to type 'ProtectFormGuard'.

67      markChanged = () => markChanged( this );
                                         ~~~~

client/lib/protect-form/index.tsx:68:32 - error TS2345: Argument of type 'this' is not assignable to parameter of type 'ProtectableForm'.
  Type 'ProtectedFormComponent' is not assignable to type 'ProtectableForm'.
    Type 'ProtectedFormComponent' is not assignable to type 'ProtectFormGuard'.
      Type 'this' is not assignable to type 'ProtectFormGuard'.
        Type 'ProtectedFormComponent' is not assignable to type 'ProtectFormGuard'.
          Types of property 'componentWillReceiveProps' are incompatible.
            Type '(nextProps: Readonly<Pick<P, SetDifference<keyof P, "markChanged" | "markSaved">>>, nextContext: any) => void' is not assignable to type '(nextProps: ProtectFormGuardProps) => void'.

68      markSaved = () => markSaved( this );
                                     ~~~~
```

Reply:
Wound up with this: https://github.com/Automattic/wp-calypso/pull/32924/files#r282674421